pages/agent/cli_pipeline.md.erb

```diff
@@ -42,13 +42,80 @@ Options:
+The pipeline can be written as YAML or JSON, but YAML is more common for its readability. There are two top level properties you can specify:
 
-* `env` - a hash of build-wide environment variables
-* `steps` - an array of build steps to be run (a mix of script, [`wait`](/docs/pipelines/wait-step) and [`block`](/docs/pipelines/block-step) steps)
+* `env` - A map of <a href="/docs/builds/environment-variables">environment variables</a> to apply to all steps
+* `steps` - A list of [build pipeline steps](/docs/pipelines/defining-steps)
 
+## Environment variable substitution
 
+In Buildkite Agent versions 3.0 and above, the `pipeline upload` command supports environment variable substitution using the syntax `$VAR` and `${VAR}`.
 
+For example, the following pipeline substitutes a number of [Buildkite’s default environment variables](/docs/builds/environment-variables) into a [trigger step](/docs/pipelines/trigger-step):
 
+```yml
+- trigger: "app-deploy"
+  label: "\:rocket\: Deploy"
+  branches: "master"
+  async: true
+  build:
+    message: "${BUILDKITE_MESSAGE}"
+    commit: "${BUILDKITE_COMMIT}"
+    branch: "${BUILDKITE_BRANCH}"
+```
+
+### Buildkite Agent 2.x support
+
+If you are unable to upgrade your agent to version 3.0 or above, it is possible (but not recommended) to emulate the agent’s environment variable substitution using bash:
+
+```shell
+echo $(eval echo $(cat pipeline.yml)) | buildkite-agent pipeline upload
+```
+
+### Handling blank and missing values
+
+If an environment variable has not been set it will evaluate to a blank string. You can set a fallback value using the syntax `${VAR:-default-value}`.
+
+For example, the following step will run the command `deploy.sh staging`:
+
+```yaml
+- command: "deploy.sh \"${SERVER:-staging}\""
+```
+
+<table>
+  <thead>
+    <tr><th>Environment Variables</th><th>Syntax</th><th>Result</th></tr>
+  </thead>
+  <tbody>
+    <tr><td><code></code></td><td><code>"${SERVER:-staging}"</code></td><td><code>"staging"</code></td></tr>
+    <tr><td><code>SERVER=""</code></td><td><code>"${SERVER:-staging}"</code></td><td><code>"staging"</code></td></tr>
+    <tr><td><code>SERVER="staging-5"</code></td><td><code>"${SERVER:-staging}"</code></td><td><code>"staging-5"</code></td></tr>
+  </tbody>
+</table>
+
+If you need to substitute environment variables containing empty strings, you can use the syntax `${VAR-default-value}` (notice the missing `:`).
+
+<table>
+  <thead>
+    <tr><th>Environment Variables</th><th>Syntax</th><th>Result</th></tr>
+  </thead>
+  <tbody>
+    <tr><td><code></code></td><td><code>"${SERVER-staging}"</code></td><td><code>"staging"</code></td></tr>
+    <tr><td><code>SERVER=""</code></td><td><code>"${SERVER-staging}"</code></td><td><code>""</code></td></tr>
+    <tr><td><code>SERVER="staging-5"</code></td><td><code>"${SERVER-staging}"</code></td><td><code>"staging-5"</code></td></tr>
+  </tbody>
+</table>
+
+### Character ranges
```

Comment:
Might wanna mention "escaping" as well:

```
steps:
  - command: "$$FOO"
  - command: "\$FOO"
```

Will result in a commands of "$FOO" being sent to the agent (including the $)